gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java

```diff
@@ -1252,20 +1261,86 @@ else if (qVal instanceof AbstractGfxdReplayableMessage) {
         }
       }
 
-      for (GemFireContainer container : uninitializedContainers) {
-        if (logger.infoEnabled() &&
-            !Misc.isSnappyHiveMetaTable(container.getSchemaName())) {
-          logger.info("FabricDatabase: start initializing container: "
-              + container);
+      // In case of reconnect, same cache is used and many locks are
+      // already taken by reconnect thread.
+      // Can't do initlization in different thread.
+      if (Thread.currentThread().getName().equals("ReconnectThread")) {
+        for (GemFireContainer container : uninitializedContainers) {
+          if (logger.infoEnabled() &&
+              !Misc.isSnappyHiveMetaTable(container.getSchemaName())) {
+            logger.info("FabricDatabase: start initializing container: "
+                + container);
+          }
+          container.initializeRegion();
+          // wait for notification
+          if (logger.infoEnabled() &&
+              !Misc.isSnappyHiveMetaTable(container.getSchemaName())) {
+            logger.info("FabricDatabase: end initializing container: "
+                + container);
+          }
         }
-        container.initializeRegion();
-        if (logger.infoEnabled() &&
-            !Misc.isSnappyHiveMetaTable(container.getSchemaName())) {
-          logger.info("FabricDatabase: end initializing container: "
-              + container);
+      }
+      else {
+        ExecutorService execService = Executors
+            .newCachedThreadPool(new ThreadFactory() {
+              AtomicInteger threadNum = new AtomicInteger(0);
+
+              public Thread newThread(final Runnable r) {
+                LogWriterI18n logger = InternalDistributedSystem.getLoggerI18n();
+                Thread result = new Thread(LogWriterImpl.createThreadGroup(
+                    "RegionInitializerExecutionThreadGroup", logger), r,
+                    "RegionInitializer Thread-" + threadNum.incrementAndGet());
+                result.setDaemon(true);
+                return result;
+              }
+            });
+
+
+        List<Callable> tasks = new ArrayList<Callable>();
+        List<Future<Object>> results = new LinkedList<>();
+        for (GemFireContainer container : uninitializedContainers) {
+          if (logger.infoEnabled() &&
+              !Misc.isSnappyHiveMetaTable(container.getSchemaName())) {
+            logger.info("FabricDatabase: start initializing container: "
+                + container);
+          }
+          // do 1 at a time
+          // check if one is done then submit next
+          final FabricService service = FabricServiceManager
+              .currentFabricServiceInstance();
+          Future f = execService.submit(new Callable<Boolean>() {
+            @Override
+            public Boolean call() throws Exception {
+              container.initializeRegion();
```

Comment:
What happens if init fails for some reason. Will it give proper error here. The Executor service won't show the reason properly as the exception will come in a different thread. Should we catch a runtime exception and log it as well.

Reply:
Will check. Generally the exception is wrapped in ExecutionException for what I remember. Do we want to strip ExecutionException?

Reply:
It should be fine.